packages/transformers/react-refresh-wrap/src/ReactRefreshWrapTransformer.js

```diff
@@ -0,0 +1,133 @@
+// @flow
+
+import semver from 'semver';
+import path from 'path';
+import {Transformer} from '@parcel/plugin';
+import {relativeUrl} from '@parcel/utils';
+import SourceMap from '@parcel/source-map';
+import generate from '@babel/generator';
+import {parse} from '@babel/parser';
+import template from '@babel/template';
+import * as t from '@babel/types';
+
+const WRAPPER = path.join(__dirname, 'helpers', 'helpers.js');
+
+const wrapper = template(`
+var helpers = require(%%helper%%);
+var prevRefreshReg = window.$RefreshReg$;
+var prevRefreshSig = window.$RefreshSig$;
+helpers.prelude(module);
+
+try {
+  %%module%%
+  helpers.postlude(module);
+} finally {
+  window.$RefreshReg$ = prevRefreshReg;
+  window.$RefreshSig$ = prevRefreshSig;
+}
+`);
+
+function shouldExclude(asset, options) {
+  return (
+    !asset.env.isBrowser() ||
+    !options.hot ||
+    !asset.isSource ||
+    !asset.getDependencies().find(v => v.moduleSpecifier === 'react')
+  );
+}
+
+export default new Transformer({
+  canReuseAST({ast}) {
+    return ast.type === 'babel' && semver.satisfies(ast.version, '^7.0.0');
+  },
+
+  async parse({asset, options}) {
+    let code = await asset.getCode();
+    if (shouldExclude(asset, options)) {
+      return null;
+    }
+
+    return {
+      type: 'babel',
+      version: '7.0.0',
+      isDirty: false,
+      program: parse(code, {
+        filename: this.name,
+        allowReturnOutsideFunction: true,
+        strictMode: false,
+        sourceType: 'module',
+        plugins: ['exportDefaultFrom', 'exportNamespaceFrom', 'dynamicImport']
+      })
+    };
+  },
+
+  async transform({asset, options}) {
+    asset.type = 'js';
+    let ast = asset.ast;
+    if (!ast || shouldExclude(asset, options)) {
+      return [asset];
+    }
+
+    asset.meta.reactRefreshRuntimePath = (await options.packageManager.resolve(
+      'react-refresh/runtime',
+      __filename
+    )).resolved;
+
+    let wrapperPath = path.relative(path.dirname(asset.filePath), WRAPPER);
+
+    ast.program.program.body = wrapper({
```

Comment:
Would this be simpler as a string transform instead of an AST one? Could possibly avoid most of the code in this transformer for parsing/generating.

Reply:
But then I'd have to patch the sourcemap?

Reply:
ugh, sourcemaps! 🙈